### PR TITLE
perf(spec): add vocab-parallel MTP local argmax

### PR DESCRIPTION
**Pull request by ywgrit**
## Motivation

DeepSeek/GLM NextN draft models use a tensor-parallel LM head. Today the greedy
`topk=1` path first all-gathers every rank's local vocabulary logits and only
then computes argmax. For this narrow mode, the worker only needs one global
token id, so communicating the full vocabulary is unnecessary.

This PR adds an **opt-in, high-batch fast path** that reduces one local maximum
candidate per rank instead. It remains disabled by default because the packed
collective has a roughly 45--48 us fixed cost on H20 and regresses small row
counts.

## What changes

- Compute each rank's local `(max_logit, global_token_id)` after excluding
  vocabulary padding.
- Pack value and id into one FP32 all-gather. Token ids are exact because the
  path is gated to `vocab_size < 2**24`.
- Select the maximum value and, on equal values, the smallest global token id,
  matching full-vocabulary `torch.argmax` tie-breaking.
- Expose the capability from `DeepseekV3ForCausalLMNextN` and consume it in all
  three proposal sites: draft-decode, prefill seed, and draft-extend.
- Add `--speculative-draft-local-argmax`, disabled by default.
- Fail closed for TP1, non-CUDA backends, `topk != 1`, rejection sampling,
  token maps, DP attention/LM-head modes, TP LM-head all-to-all, added
  vocabulary, LoRA, unsupported heads, and oversized vocabularies.

## Runtime flow

```mermaid
flowchart TD
    A["EagleDraftWorker proposal phase"] --> B["DeepseekV3ForCausalLMNextN.forward"]
    B --> C["ParallelLMHead produces local-vocab logits"]
    C --> D{"opt-in eligibility gate"}
    D -->|"disabled or unsupported"| E["existing full-vocab AllGather"]
    E --> F["existing top-1 postprocess"]
    D -->|"greedy TP fast path"| G["local max and global token id"]
    G --> H["one packed FP32 candidate AllGather"]
    H --> I["deterministic global max with first-id tie-break"]
    F --> J["draft token and existing bookkeeping"]
    I --> J
    class G,H,I changed;
    classDef changed stroke-dasharray:5 5,stroke-width:2px;
```

The model decides at construction time whether it may skip the full-logits
gather. The worker then uses the same capability in every place that consumes
the draft logits. Unsupported configurations retain the existing full-logits
path.

## Correctness and validation

GPU/runtime validation was collected on
`3e61fa357e8cf82244d305f1e25f6e7886ca924b`. The current PR head
`939ffe8240f63e55c9b993521e731edb80284600` differs only by repository-required
Black formatting and `register_cpu_ci` metadata; the runtime behavior is
unchanged.

- Focused suite: **27 passed** in a compatible SGLang environment.
- TP2 full-vocabulary parity for FP16/BF16/FP32, including cross-rank ties,
  padded and empty shards, shard offsets, and random inputs.
- Exactly one packed collective per proposal.
- Draft-decode, prefill-seed, and draft-extend bookkeeping coverage.
- Production CUDA Graph capture/replay parity and stable tensor shapes.
- Fail-closed matrix for every unsupported mode listed above.
- Fresh pre-submission checks: `compileall`, Ruff `F401/F821/UP037`, and
  `git diff --check` pass.

The local laptop's PyTorch 2.4.1 cannot collect current SGLang because it lacks
`torch.distributed._symmetric_memory`; the 27-test result above is from the same
source SHA on the H20 SGLang environment.

## H20 performance

Environment: 2 x H20 96 GB, CUDA 13, BF16 logits, vocabulary 151,552, CUDA
Graph steady state, 20 warmups, 100 operations per repeat, 10 interleaved A/B
repeats. Reported latency is the median of the slowest TP rank.

| Rows | Full-vocab baseline | Packed candidate | Speedup |
|---:|---:|---:|---:|
| 1 | 15.072 us | 43.472 us | 0.347x |
| 8 | 23.371 us | 44.977 us | 0.520x |
| 16 | 34.080 us | 44.720 us | 0.762x |
| 32 | 55.258 us | 45.074 us | **1.226x** |
| 64 | 93.586 us | 45.628 us | **2.051x** |
| 128 | 185.442 us | 48.470 us | **3.826x** |

At 128 rows, each rank's logical receive payload falls from 19,398,656 bytes
to 1,024 bytes (**18,944x smaller**). Nsight Systems confirms that the fast
path replaces the full-vocabulary gather/argmax boundary with the packed
candidate collective.

These are TP operator results, not full GLM-5.2 serving results. The PR does not
claim an end-to-end throughput gain, and the feature remains off by default
until representative production workloads justify a default heuristic.

## Relationship to nearby work

- #26235 skips full-vocabulary softmax after logits have already been gathered.
- #30947 fuses the existing top-1 postprocess on full logits.
- Open #31465 extends that postprocess to draft-extend selected rows.

This PR changes the earlier LM-head communication boundary: it prevents the
full-vocabulary TP gather in the first place. It reuses the existing worker
bookkeeping and does not replace the postprocess kernels in those PRs.

## Review focus

I would especially appreciate feedback on whether the construction-time
capability contract is the right integration boundary, and whether maintainers
prefer to keep this as an explicit experimental flag or add a backend-specific
high-batch dispatch heuristic later.

### CODE REVIEW

**Comment by ywgrit** [comment]
Hi @Qiaolin-Yu and @mattteochen, could you please review this focused MTP/spec-decoding change? Your #26235, #30947, and #31465 work defines the closest existing top-1 postprocess paths; this PR changes an earlier boundary by computing greedy argmax over sharded logits with one packed FP32 (max, id) collective instead of gathering the full vocabulary.

The current one-commit head passed official lint and local focused tests. Its runtime code was validated on 2×H20 for bit-exact output, ties, padding, empty shards, single-collective bookkeeping, CUDA Graph replay, and Nsight. The high-row Graph cases are 1.226×/2.051×/3.826× at rows 32/64/128; rows 1/8/16 are slower and therefore fail closed to the existing path. The feature is default-off and the PR does not claim full GLM-5.2 E2E gains.

I would especially appreciate feedback on the packed reduction semantics and whether the eligibility boundary belongs in the NextN model path.
